go/api/config.go

```diff
@@ -0,0 +1,309 @@
+// Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
+
+package api
+
+import "github.com/aws/glide-for-redis/go/glide/protobuf"
+
+const (
+	defaultHost = "localhost"
+	defaultPort = 6379
+)
+
+// NodeAddress represents the host address and port of a node in the cluster.
+type NodeAddress struct {
+	Host string // Optional: if not supplied, "localhost" will be used.
+	Port int    // Optional: if not supplied, 6379 will be used.
+}
+
+func (addr *NodeAddress) toProtobuf() *protobuf.NodeAddress {
+	if addr.Host == "" {
+		addr.Host = defaultHost
+	}
+
+	if addr.Port == 0 {
+		addr.Port = defaultPort
+	}
+
+	return &protobuf.NodeAddress{Host: addr.Host, Port: uint32(addr.Port)}
+}
+
+// RedisCredentials represents the credentials for connecting to a Redis server.
+type RedisCredentials struct {
+	// Optional: the username that will be used for authenticating connections to the Redis servers. If not supplied, "default"
+	// will be used.
+	username string
+	// Required: the password that will be used for authenticating connections to the Redis servers.
+	password string
+}
+
+// NewRedisCredentials returns a new RedisCredentials struct with the given username and password.
+func NewRedisCredentials(username string, password string) *RedisCredentials {
+	return &RedisCredentials{username, password}
+}
+
+// NewRedisCredentialsWithDefaultUsername returns a new RedisCredentials struct with a default username of "default" and the
+// given password.
+func NewRedisCredentialsWithDefaultUsername(password string) *RedisCredentials {
+	return &RedisCredentials{password: password}
+}
+
+func (creds *RedisCredentials) toProtobuf() *protobuf.AuthenticationInfo {
+	return &protobuf.AuthenticationInfo{Username: creds.username, Password: creds.password}
```

Comment:
This PR does not have the required logic to connect, it only adds code for the client configurations and the logic to convert them to protobuf connection requests, so we can't get the Go logs for the client yet. I agree that we should verify this once the connection logic is in place. 

Note that the Go files generated by protoc from the *.proto files have string and uint32 fields. In Go, string and uint32 variables cannot be given a nil value. This means that unset string and uint32 fields default to "" and 0 instead of `nil`. In this PR, the protobuf fields are only set to non-defaults if the user has set a non-default value on the config (eg a non-empty string or non-zero int). On this particular line, if the user has not set the username, the protobuf username will be still have the default unset value of "".

Reply:
please verify this once you have actual connections.

Reply:
Sounds good.

Reply:
I'm currently working on a branch that has the connect logic. The branch isn't ready for a PR yet but I was able to connect and view the logs. I tried with Python and Go using a config that only has the address set. In both cases I see this same log:
```
Connection configuration - 
Addresses: localhost:6379
TLS mode: No TLS
Standalone mode
Read from Replica mode: Only primary
Protocol: RESP3
```